src/zql/context/replicache-context.ts

```diff
@@ -0,0 +1,147 @@
+import {ExperimentalNoIndexDiff, Replicache} from 'replicache';
+import {Materialite} from '../ivm/materialite.js';
+import {Entity} from '../../generate.js';
+import {Source} from '../ivm/source/source.js';
+import {MutableSetSource} from '../ivm/source/set-source.js';
+import {assert, invariant} from '../error/asserts.js';
+import {compareEntityFields} from '../ivm/compare.js';
+import {Ordering} from '../ast/ast.js';
+
+export function makeReplicacheContext(rep: Replicache) {
+  const materialite = new Materialite();
+  const sourceStore = new ReplicacheSourceStore(rep, materialite);
+
+  return {
+    materialite,
+    getSource: <T extends Entity>(
+      name: string,
+      ordering?: [string[], 'asc' | 'desc'],
+    ) => sourceStore.getSource(name, ordering) as Source<T>,
+  };
+}
+
+/**
+ * Forwards Replicache changes to Materialite sources so they
+ * can be fed into any queries that may exist.
+ *
+ * Maintains derived orderings of sources as well.
+ *
+ * If someone runs a query that has an order-by we need to scan the entire collection
+ * in order to sort it.
+ * To save future work, we save the result of that sort and keep it up to date.
+ *
+ * This helps:
+ * 1. When revisting old queries that were sorted or paging through results
+ * 2. When many queries are sorted by the same field
+ * 3. When joining sources on a field that we have pre-sorted
+ *
+ * And shares the work between queries.
+ */
+class ReplicacheSourceStore {
+  readonly #rep: Replicache;
+  readonly #materialite: Materialite;
+  readonly #sources = new Map<string, ReplicacheSource>();
+
+  constructor(rep: Replicache, materialite: Materialite) {
+    this.#rep = rep;
+    this.#materialite = materialite;
+  }
+
+  getSource(name: string, ordering?: Ordering) {
+    let source = this.#sources.get(name);
+    if (source === undefined) {
+      source = new ReplicacheSource(this.#rep, this.#materialite, name);
+      this.#sources.set(name, source);
+    }
+
+    return source.get(ordering);
+  }
+}
+
+class ReplicacheSource {
+  readonly #materialite;
+  readonly #sources: Map<string, Source<Entity>> = new Map();
+  readonly #canonicalSource: MutableSetSource<Entity>;
+
+  constructor(rep: Replicache, materialite: Materialite, name: string) {
+    this.#canonicalSource =
+      materialite.newSetSource<Entity>(canonicalComparator);
+    this.#materialite = materialite;
+    rep.experimentalWatch(this.#onReplicacheDiff, {
+      prefix: `${name}/`,
+      initialValuesInFirstDiff: true,
+    });
+  }
+
+  #onReplicacheDiff = (changes: ExperimentalNoIndexDiff) => {
+    this.#materialite.tx(() => {
```

Comment:
I'm wondering if the materialite.tx should be around all of the diffs (for all of the source tables) for an underlying Replicache commit.   I think achieving this would require a deeper integration with Replicache.

Reply:
Well actually since we dispatch all the diffs synchronously, I think we can achieve it without deeper integration, if we have a version of materialite.tx with explicit materialite.txStart/materialite.txEnd.  If a tx is not open #onReplicacheDiff, we do materialite.txStart, and enqueue a microtask to do materialite.txEnd.

Reply:
> I'm wondering if the materialite.tx should be around all of the diffs (for all of the source tables) for an underlying Replicache commit. 

yep, it should. Good catch.

Reply:
Good catch. That might also explain some odd behaviors I'm seeing.